cardano-api/src/Cardano/Api/Governance/Poll.hs

```diff
@@ -0,0 +1,391 @@
+import           Cardano.Api.SerialiseCBOR
+import           Cardano.Api.SerialiseRaw
+import           Cardano.Api.SerialiseTextEnvelope
+import           Cardano.Api.SerialiseUsing
+import           Cardano.Api.TxMetadata
+import           Cardano.Api.Utils
+
+import           Cardano.Binary (DecoderError(..))
+import           Cardano.Ledger.Crypto (HASH, StandardCrypto, VRF)
+import           Cardano.Ledger.Keys (KeyRole(..), SignedDSIGN, SignKeyDSIGN,
+                   SignKeyVRF, VKey(..), VerKeyVRF, signedDSIGN, verifySignedDSIGN)
+
+import qualified Cardano.Crypto.DSIGN as DSIGN
+import           Cardano.Crypto.Hash (hashFromBytes, hashToBytes, hashWith)
+import qualified Cardano.Crypto.Hash as Hash
+import           Cardano.Crypto.Util (SignableRepresentation(..))
+import qualified Cardano.Crypto.VRF as VRF
+
+-- | Associated metadata label as defined in CIP-XXXX
+metadataLabel :: Word64
+metadataLabel = 94
+
+-- ----------------------------------------------------------------------------
+-- Governance Poll
+--
+
+data GovernancePoll = GovernancePoll
+    { govPollQuestion :: Text
+    , govPollAnswers :: [Text]
+    , govPollNonce :: Maybe Word
+    }
+  deriving Show
+
+instance HasTextEnvelope GovernancePoll where
+   textEnvelopeType _ = "GovernancePoll"
+
+instance HasTypeProxy GovernancePoll where
+    data AsType GovernancePoll = AsGovernancePoll
+    proxyToAsType _ = AsGovernancePoll
+
+instance AsTxMetadata GovernancePoll where
+    asTxMetadata GovernancePoll{govPollQuestion, govPollAnswers, govPollNonce} =
+      makeTransactionMetadata $ Map.fromList
+        [ ( metadataLabel
+          , TxMetaMap $
+           [ ( TxMetaNumber 0, TxMetaText govPollQuestion )
+           , ( TxMetaNumber 1, TxMetaList (TxMetaText <$> govPollAnswers) )
+           ] ++
+           case govPollNonce of
+             Nothing -> []
+             Just nonce -> [ ( TxMetaText "_", TxMetaNumber (toInteger nonce) ) ]
+          )
+        ]
+
+instance SerialiseAsCBOR GovernancePoll where
+    serialiseToCBOR =
+      serialiseToCBOR . asTxMetadata
+
+    deserialiseFromCBOR AsGovernancePoll bs = do
+      metadata <- deserialiseFromCBOR AsTxMetadata bs
+      withNestedMap lbl metadataLabel metadata $ \values ->
+        GovernancePoll
+          <$> case lookup (TxMetaNumber 0) values of
+                Just x  -> expectText (field "0") x
+                Nothing -> Left $ missingField (field "0")
+          <*> case lookup (TxMetaNumber 1) values of
+                Just (TxMetaList xs) -> traverse (expectText (field "1")) xs
+                Just _  -> Left $ malformedField (field "1") "List of Text"
+                Nothing -> Left $ missingField (field "1")
+          <*> case lookup (TxMetaText "_") values of
+                Just (TxMetaNumber nonce) -> Just <$> expectWord (field "_") nonce
+                Nothing -> pure Nothing
+                Just _  -> Left $ malformedField (field "_") "Number (uint)"
```

Comment:
Why `uint`?

Reply:
`unsigned integer` as the rest of the world call them; Haskell folks calls them `word` 🙃